src/pynxtools/dataconverter/helpers.py

```diff
@@ -723,78 +752,91 @@ def convert_int_to_float(value):
         return {convert_int_to_float(v) for v in value}
     elif isinstance(value, np.ndarray) and np.issubdtype(value.dtype, np.integer):
         return value.astype(float)
+    elif isinstance(value, np.generic) and np.issubdtype(type(value), np.integer):
+        return float(value)
     else:
         return value
 
 
 def is_valid_data_field(
     value: Any, nxdl_type: str, nxdl_enum: list, nxdl_enum_open: bool, path: str
 ) -> Any:
-    # todo: Check this function and write test for it. It seems the function is not
-    # working as expected.
-    """Checks whether a given value is valid according to the type defined in the NXDL.
-
-    This function only tries to convert boolean value in str format (e.g. "true" ) to
-    python Boolean (True). In case, it fails to convert, it raises an Exception.
+    """Checks whether a given value is valid according to the type defined in the NXDL."""
+
+    def validate_data_value(
+        value: Any, nxdl_type: str, nxdl_enum: list, nxdl_enum_open: bool, path: str
+    ) -> Any:
+        """Validate and possibly convert a primitive value according to NXDL type/enum rules."""
+        accepted_types = NEXUS_TO_PYTHON_DATA_TYPES[nxdl_type]
+        original_value = value
+
+        # Do not count other dicts as they represent a link value
+        if not isinstance(value, dict):
+            # Attempt type conversion
+            if accepted_types[0] is bool and isinstance(value, str):
+                try:
+                    value = convert_str_to_bool_safe(value)
+                except (ValueError, TypeError):
+                    value = original_value
+            elif accepted_types[0] is float:
+                value = convert_int_to_float(value)
```

Comment:
Please add a warning between 781 and 782 that this conversion can be result in precision loss but keep line 782 for now, as mentioned in another comment, yeah keep it here and we should at some point make a follow-up PR which removes this, main reason, avoiding side effects now for other users that rely on this conversion.

Reply:
Consistent with my wish to limit the scope of this PR and move on this is a reasonable compromise.
We keep the conversion int to float as it is for now and should address this is in a follow-up.

My view on the is-info-or-warning is: When not every int has exactly only one equivalent float it needs a warning cuz we cannot foresee if there are really no applications that can ignore the precision loss.